src/Datadog.Trace/SpanContextPropagator.cs

```diff
@@ -60,6 +62,39 @@ public void Inject<T>(SpanContext context, T headers)
                 samplingPriority?.ToString(InvariantCulture));
         }
 
+        /// <summary>
+        /// Propagates the specified context by adding new headers to a <see cref="IHeadersCollection"/>.
+        /// This locks the sampling priority for <paramref name="context"/>.
+        /// </summary>
+        /// <param name="context">A <see cref="SpanContext"/> value that will be propagated into <paramref name="carrier"/>.</param>
+        /// <param name="carrier">The headers to add to.</param>
+        /// <param name="setter">The action that can set a header in the carrier.</param>
+        /// <typeparam name="T">Type of header collection</typeparam>
+        public void Inject<T>(SpanContext context, T carrier, Action<T, string, string> setter)
+        {
+            if (context == null) { throw new ArgumentNullException(nameof(context)); }
+
+            if (carrier == null) { throw new ArgumentNullException(nameof(carrier)); }
+
+            if (setter == null) { throw new ArgumentNullException(nameof(setter)); }
+
+            // lock sampling priority when span propagates.
+            context.TraceContext?.LockSamplingPriority();
+
+            setter(carrier, HttpHeaderNames.TraceId, context.TraceId.ToString(InvariantCulture));
```

Comment:
NIT: It feels weird to convert those to string and then have the setter convert them to binary. maybe the setter could be an `Action<T, string, byte[]>` and have `Inject` deal with the encoding

Reply:
Yeah I think that idea should work. Let me try that. In the end, I think we will also want to support an Inject that takes a `Action<T, string, string>` to support text-based (string-based) headers like HTTP headers.

Reply:
Hmm on second thought, can we leave as is for now? RabbitMQ is the one exception that needs to translate to bytes, whereas others just carry strings, but it just so happens to be the first to use this delegate-based design.